editoast/src/views/timetable/stdcm.rs

```diff
@@ -191,6 +184,43 @@ async fn stdcm(
     let earliest_departure_time = stdcm_request.get_earliest_departure_time(simulation_run_time);
     let latest_simulation_end = stdcm_request.get_latest_simulation_end(simulation_run_time);
 
+    let timetable = Timetable::retrieve_or_fail(&mut conn, timetable_id, || {
+        StdcmError::TimetableNotFound { timetable_id }
+    })
+    .await?;
+
+    // Filtering trains is done in 2 stages:
+    // 1) We filter all trains whose departure dates are before the latest_simulation_end via sql
+
+    let mut train_schedules = timetable
+        .schedules_in_time_window(&mut conn, latest_simulation_end)
+        .await?;
+
+    // 2) We then filter in rust according to the lmr train departure date.  Since TrainSchedule input is light, this rust filtering saves time.
+    train_schedules.retain(|train_schedule| {
+        if train_schedule.start_time >= earliest_departure_time {
+            true
+        } else {
+            if let Some(last_schedule_item) = train_schedule.schedule.last() {
+                if let Some(last_path_item) = train_schedule.path.last() {
+                    if last_schedule_item.at == last_path_item.id {
+                        if let Some(arrival) = last_schedule_item.arrival.clone() {
+                            train_schedule.start_time + *arrival > earliest_departure_time
+                        } else {
+                            true
+                        }
+                    } else {
+                        true
+                    }
+                } else {
+                    true
+                }
+            } else {
+                false
+            }
+        }
```

Comment:
Could this be rewritten in a more monadic style to avoid this much nesting? 😇 

```rust
 } else {
                            true
                        }
                    } else {
                        true
                    }
                } else {
                    true
                }
            } else {
                false
            }
        }
```

![image](https://github.com/user-attachments/assets/1c9d8b4a-47c5-4468-90b1-bebf695dc06d)